poly_borders/borders_data.cpp

```diff
@@ -0,0 +1,529 @@
+#include "poly_borders/borders_data.hpp"
+
+#include "poly_borders/help_structures.hpp"
+
+#include "generator/borders.hpp"
+#include "generator/routing_city_boundaries_processor.hpp"
+
+#include "platform/platform.hpp"
+
+#include "geometry/mercator.hpp"
+#include "geometry/point2d.hpp"
+#include "geometry/rect2d.hpp"
+#include "geometry/region2d.hpp"
+
+#include "base/assert.hpp"
+#include "base/file_name_utils.hpp"
+#include "base/logging.hpp"
+#include "base/scope_guard.hpp"
+#include "base/string_utils.hpp"
+#include "base/thread_pool_computational.hpp"
+#include "base/timer.hpp"
+
+#include <algorithm>
+#include <fstream>
+#include <future>
+#include <iostream>
+#include <iterator>
+#include <thread>
+#include <tuple>
+#include <utility>
+
+namespace
+{
+using namespace poly_borders;
+
+void PrintWithSpaces(std::string const & str, size_t maxN)
+{
+  std::cout << str;
+  if (maxN <= str.size())
+    return;
+
+  maxN -= str.size();
+  for (size_t i = 0; i < maxN; ++i)
+    std::cout << " ";
+}
+
+std::string RemoveIndexFromMwmName(std::string const & mwmName)
+{
+  auto const pos = mwmName.find(BordersData::kBorderExtension);
+  auto const end = mwmName.begin() + pos + BordersData::kBorderExtension.size();
+  std::string result(mwmName.cbegin(), end);
+  return result;
+}
+
+bool IsReversedIntervals(size_t fromDst, size_t toDst, size_t fromSrc, size_t toSrc)
+{
+  return (fromDst > toDst) != (fromSrc > toSrc);
+}
+
+std::vector<m2::PointD> CopyPointsWithAnyDirection(std::vector<MarkedPoint> const & copyFrom, 
+                                                   size_t from, size_t to)
+{
+  std::vector<m2::PointD> result;
+  if (from > to)
+    std::swap(from, to);
+
+  for (size_t k = from; k <= to; ++k)
+    result.emplace_back(copyFrom[k].m_point);
+  
+  return result;
+}
+
+double AbsAreaDiff(std::vector<m2::PointD> const & firstPolygon,
+                   std::vector<m2::PointD> const & secondPolygon)
+{
+  auto const currentArea = generator::routing_city_boundaries::AreaOnEarth(firstPolygon);
+  auto const anotherArea = generator::routing_city_boundaries::AreaOnEarth(secondPolygon);
+  return std::abs(anotherArea - currentArea);
+}
+
+bool NeedReplace(std::vector<m2::PointD> const & curSubpolygon,
+                 std::vector<m2::PointD> const & anotherSubpolygon)
+{
+  auto const areaDiff = AbsAreaDiff(curSubpolygon, anotherSubpolygon);
+  double constexpr kMaxAreaDiff = 20000.0;
+  if (areaDiff > kMaxAreaDiff)
+    return false;
+
+  if (base::AlmostEqualAbs(areaDiff, 0.0, BordersData::kEqualityEpsilon))
+    return false;
+
+  // We know that |curSize| is always greater than 1, because we construct it such way, but we know
+  // nothing about |anotherSize|, and we do not want to replace current subpolygon of several points
+  // to subpolygon consists of one point.
+  CHECK_GREATER(curSubpolygon.size(), 1, ());
+  return anotherSubpolygon.size() > 1 &&
+         std::max(curSubpolygon.size(), anotherSubpolygon.size()) < 10;
```

Comment:
```cpp
// We know that |curSize| is always greater than 1, because we construct it such way, but we know
  // nothing about |anotherSize|, and we do not want to replace current subpolygon of several points
  // with a subpolygon consisting of one point or consisting of too many points.
  CHECK_GREATER(curSubpolygon.size(), 1, ());
  return 1 < anotherSubpolygon.size() && anotherSubpolygon.size() < 10;
```
так понятнее?